### PR TITLE
fix(npm): use ntfs junctions in node_modules folder on Windows

### LINKED ISSUE #16006: Local node_modules - Use junctions instead of symlinks on Windows
Edit: Use junctions instead like pnpm does (see comments below)

~~Users on Windows need to enable developer mode to get the local node_modules working because it uses symlinks.~~

https://learn.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development

~~I think though that some users might not be able to enable this, but I'm not sure how widespread of an issue this is.. so for now, let's add some instructions about enabling developer mode then maybe in the future have a warning then fallback solution for those users.~~

---

See error from: https://github.com/denoland/deno/issues/15427#issuecomment-1255657707

```
PS C:\Users\rekke\Documents\vite-deno-example> deno task dev
Warning deno task is unstable and may drastically change in the future
Task dev deno run -A --unstable --node-modules-dir npm:vite
error: A required privilege is not held by the client. (os error 1314), symlink 'C:\Users\rekke\Documents\vite-deno-example\node_modules\.deno\esbuild-freebsd-arm64@0.15.8\node_modules\esbuild-freebsd-arm64' -> 'C:\Users\rekke\Documents\vite-deno-example\node_modules\.deno\esbuild@0.15.8\node_modules\esbuild-freebsd-arm64'
```

### PR DESCRIPTION
**Pull request by dsherret**
Does what pnpm does (https://pnpm.io/faq#does-it-work-on-windows) and what was recommended in the issue.

Closes #16006

### CODE REVIEW

File: cli/npm/resolvers/local.rs [hunk 1]
```diff
@@ -323,9 +323,39 @@ fn symlink_package_dir(
...
+  match junction::create(old_path, new_path) {
+    Ok(()) => Ok(()),
+    Err(junction_err) => {
+      match fs_util::symlink_dir(old_path, new_path) {
+        Ok(()) => Ok(()),
+        Err(symlink_err) => bail!(
+          concat!(
+            "Attempted to junction and failed. Then attempted to symlink and also failed.\n\n",
+            "Junction error:\n{:#}\n\nSymlink error:\n{:#}\n\n",
+            "Maybe try enabling developer mode on your machine?"
```
Comment (cjihrig) [inline cli/npm/resolvers/local.rs:349]: Not the most reassuring error message 😄 Can we use more direct wording?

**Review by bartlomieju — APPROVED** [review]
LGTM

File: cli/npm/resolvers/local.rs [hunk 2]
```diff
@@ -323,9 +323,39 @@ fn symlink_package_dir(
 
   // need to delete the previous symlink before creating a new one
   let _ignore = fs::remove_dir_all(new_path);
+
+  #[cfg(windows)]
+  return junction_or_symlink_dir(old_path, new_path);
```
Comment (cjihrig) [inline cli/npm/resolvers/local.rs:328]: For consistency, would it work to use an implicit return similar to the `not(windows)` path (if that doesn't work in Rust and I'm way off, sorry).

Comment (dsherret) [reply cli/npm/resolvers/local.rs:328]: Yeah, it doesn't compile. I'm not sure why.

File: cli/npm/resolvers/local.rs (hunk 1 above)
Comment (dsherret) [reply cli/npm/resolvers/local.rs]: I tweaked it a bit. Let me know if you have any suggestions.

Comment (cjihrig) [reply cli/npm/resolvers/local.rs]: I was thinking of something along the lines of replacing:

"Maybe try enabling developer mode on your machine?"

with:

"Please try enabling developer mode on your machine. If you continue to experience this error, please open an issue."

Comment (dsherret) [reply cli/npm/resolvers/local.rs]: I'm just going to remove this line because it actually might not be the right solution.

**Pull request merged**